File: src/qaeval/openllm.js
// openllm.js - Ollama API를 사용한 LLM 평가 모듈
import fs from 'fs/promises';
import path from 'path';
import { fileURLToPath } from 'url';
import fetch from 'node-fetch';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// 전역 설정
const DEFAULT_CONFIG = {
    model: 'mistral',
    promptPath: path.join(__dirname, '../../prompts/eval-v0.1-zero-shot.txt'),
    candidatesPath: path.join(__dirname, '../../data/candidates.tsv')
};

/**
 * TSV 파일에서 candidates를 읽어오는 함수
 * @param {string} filePath - TSV 파일 경로
 * @returns {Promise<Array>} candidates 배열
 */
async function readCandidatesFromTSV(filePath) {
    try {
        const content = await fs.readFile(filePath, 'utf-8');
        const lines = content.trim().split('\n');

        // 헤더 제거
        const headers = lines[0].split('\t');
        const data = lines.slice(1);

        return data.map(line => {
            const [question, answers, candidate_answer] = line.split('\t');
            return {
                question,
                answers: answers.split(',').map(a => a.trim()),
                candidate_answer
            };
        });
    } catch (error) {
        logger.error(`TSV 파일 읽기 실패: ${error.message}`);
        throw error;
    }
}

// command line argument 파싱
function parseArgs() {
    const args = process.argv.slice(2);
    const config = { ...DEFAULT_CONFIG };

    for (let i = 0; i < args.length; i++) {
        const arg = args[i];
        if (arg === '--model' && args[i + 1]) {
            config.model = args[++i];
        } else if (arg === '--prompt' && args[i + 1]) {
            config.promptPath = args[++i];
        } else if (arg === '--candidates' && args[i + 1]) {
            config.candidatesPath = args[++i];
        } else if (arg === '--help') {
            console.log(`
사용법: node openllm.js [옵션]

옵션:
  --model <모델명>        Ollama 모델명 (기본값: ${DEFAULT_CONFIG.model})
  --prompt <경로>         프롬프트 파일 경로 (기본값: ${DEFAULT_CONFIG.promptPath})
  --candidates <경로>     후보 답변 TSV 파일 경로 (기본값: ${DEFAULT_CONFIG.candidatesPath})
  --help                 도움말 표시

TSV 파일 형식:
  question<TAB>answers<TAB>candidate_answer
  answers는 콤마(,)로 구분된 문자열
예시:
  What is the capital of France?<TAB>Paris,Paris City<TAB>Paris
            `);
            process.exit(0);
        }
    }

    return config;
}

// 로깅 설정
const logger = {
    info: (...args) => console.log('[INFO]', ...args),
    warning: (...args) => console.warn('[WARNING]', ...args),
    error: (...args) => console.error('[ERROR]', ...args)
};

/**
 * Ollama API를 통해 프롬프트를 전송하고 응답을 받는 함수
 * @param {string} prompt - 전송할 프롬프트
 * @param {string} modelName - Ollama 모델명 (예: 'mistral', 'zephyr:7b', 'vicuna:13b')
 * @returns {Promise<string>} 모델의 응답
 */
async function sendToOllama(prompt, modelName) {
    try {
        logger.info('=== Ollama API 호출 ===');
        logger.info('모델:', modelName);
        logger.info('프롬프트:', prompt);

        const response = await fetch('http://localhost:11434/api/generate', {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({
                model: modelName,
                prompt: prompt,
                stream: false
            })
        });

        if (!response.ok) {
            throw new Error(`HTTP error! status: ${response.status}`);
        }

        const result = await response.json();
        logger.info('응답:', result.response);
        logger.info('===================');
        return result.response || '';
    } catch (error) {
        logger.error(`Ollama API 호출 실패: ${error.message}`);
        return '';
    }
}

/**
 * 프롬프트 템플릿을 준비하는 함수
 * @param {Object} candidate - 평가할 후보 답변 객체
 * @param {string} promptTemplate - 프롬프트 템플릿
 * @returns {string} 완성된 프롬프트
 */
function preparePrompt(candidate, promptTemplate) {
    const { question, answers, candidate_answer } = candidate;
    const gold_answers = answers.join(', ');

    return promptTemplate
        .replace('{q}', question)
        .replace('{answers}', gold_answers)
        .replace('{candidate_answer}', candidate_answer);
}

/**
 * 모델의 응답을 파싱하여 yes/no 판단을 반환하는 함수
 * @param {string} response - 모델의 응답
 * @param {string} candidateAnswer - 후보 답변
 * @param {string} question - 질문
 * @returns {number} 1(yes) 또는 0(no)
 */
function parseResponse(response, candidateAnswer, question) {
    // 응답 전처리: 앞뒤 공백 제거 및 소문자 변환
    const cleanResponse = response.trim().toLowerCase();

    // "yes" 또는 "no"로 시작하는 경우
    if (cleanResponse.startsWith('yes')) {
        return 1;
    }
    if (cleanResponse.startsWith('no')) {
        return 0;
    }

    // "yes" 또는 "no"를 포함하는 패턴들
    const patterns = [
        /.*['"]?(yes|no)\.?['"]?[.!]?$/i,
        /.*I can answer\s+['"]?(yes|no)['"]?[.!]?/i,
        /.*I would say\s+['"]?(yes|no)['"]?[.!]?/i,
        /.*I must say\s+['"]?(yes|no)['"]?[.!]?/i,
        /.*my (final )?judgment is\s+['"]?(yes|no)['"]?[.!]?/i,
        /.*I would judge the candidate answer as\s+['"]?(yes|no)['"]?[.!]?/i,
        /.*\s+['"]?(yes|no)['"]?,? the candidate( answer)? is/i,
        /.*[jJ]udgment:\s+['"]?(yes|no)\.?['"]?/i,
        /.*the candidate( answer)? is (correct|incorrect)/i,
        /.*the answer is (correct|incorrect)/i
    ];

    // 긍정적인 표현을 포함하는 패턴들
    const positivePatterns = [
        /candidate( answer)? is correct/i,
        /candidate's correct/i,
        /answer is correct/i,
        /correct answer/i,
        /matches the ground-truth/i,
        /is accurate/i,
        /is valid/i
    ];

    // 부정적인 표현을 포함하는 패턴들
    const negativePatterns = [
        /candidate( answer)? is incorrect/i,
        /candidate's incorrect/i,
        /answer is incorrect/i,
        /incorrect answer/i,
        /does not match/i,
        /is inaccurate/i,
        /is invalid/i
    ];

    // 패턴 매칭 시도
    for (const pattern of patterns) {
        const match = cleanResponse.match(pattern);
        if (match) {
            const answer = match[1]?.toLowerCase();
            if (answer === 'yes' || answer === 'correct') return 1;
            if (answer === 'no' || answer === 'incorrect') return 0;
        }
    }

    // 긍정/부정 패턴 확인
    for (const pattern of positivePatterns) {
        if (pattern.test(cleanResponse)) {
            return 1;
        }
    }

    for (const pattern of negativePatterns) {
        if (pattern.test(cleanResponse)) {
            return 0;
        }
    }

    // 응답에 "yes" 또는 "no"가 포함되어 있는지 확인
    if (cleanResponse.includes('yes')) return 1;
    if (cleanResponse.includes('no')) return 0;

    logger.warning(`Invalid response to \`${question}\` & \`${candidateAnswer}\`: ${response}`);
    return 0;
}

/**
 * LLM 평가를 실행하는 메인 함수
 * @param {string} modelName - Ollama 모델명
 * @param {Array} candidates - 평가할 후보 답변 배열
 * @param {string} promptFile - 프롬프트 템플릿 파일 경로
 * @returns {Promise<Array>} 평가 결과 배열
 */
export async function llmEval(modelName, candidates, promptFile) {
    try {
        // 프롬프트 템플릿 읽기
        const promptTemplate = await fs.readFile(promptFile, 'utf-8');
        logger.info('=== 프롬프트 템플릿 ===');
        logger.info(promptTemplate);
        logger.info('===================');

        // 각 후보에 대해 평가 수행
        const results = [];
        for (const candidate of candidates) {
            logger.info('=== 평가 항목 ===');
            logger.info('질문:', candidate.question);
            logger.info('정답:', candidate.answers);
            logger.info('후보 답변:', candidate.candidate_answer);

            const prompt = preparePrompt(candidate, promptTemplate);
            const response = await sendToOllama(prompt, modelName);
            const score = parseResponse(response, candidate.candidate_answer, candidate.question);
            results.push(score);

            logger.info('점수:', score);
            logger.info('===================');
        }

        return results;
    } catch (error) {
        logger.error(`평가 중 오류 발생: ${error.message}`);
        throw error;
    }
}

// 테스트 실행 함수
async function runTest() {
    try {
        const config = parseArgs();

        // 프롬프트 파일 존재 확인
        try {
            await fs.access(config.promptPath);
        } catch (error) {
            logger.error(`프롬프트 파일을 찾을 수 없습니다: ${config.promptPath}`);
            return;
        }

        // candidates 파일 존재 확인
        try {
            await fs.access(config.candidatesPath);
        } catch (error) {
            logger.error(`candidates 파일을 찾을 수 없습니다: ${config.candidatesPath}`);
            return;
        }

        logger.info('=== 설정 ===');
        logger.info('모델:', config.model);
        logger.info('프롬프트:', config.promptPath);
        logger.info('Candidates:', config.candidatesPath);
        logger.info('===================');

        // candidates 읽기
        const candidates = await readCandidatesFromTSV(config.candidatesPath);
        logger.info(`총 ${candidates.length}개의 후보 답변을 읽었습니다.`);

        logger.info('평가 시작...');
        const results = await llmEval(config.model, candidates, config.promptPath);
        logger.info('평가 결과:', results);
    } catch (error) {
        logger.error('테스트 실행 중 오류:', error);
    }
}

// 테스트 실행
runTest();

File: src/qaeval/openllm.py
import json
import logging
import os
import re
from typing import Optional, Sequence, Union
from tqdm import tqdm

import datasets
import ftfy
import torch
from torch.utils.data import DataLoader
from transformers import (
    AutoConfig,
    AutoModelForCausalLM,
    AutoModelForSeq2SeqLM,
    AutoTokenizer,
    DataCollatorWithPadding,
    DataCollatorForSeq2Seq,
)
import requests

from .data_utils import Candidate

logger = logging.getLogger("openllm")


CONVERSATIONAL_MODELS = {
    "meta-llama/Llama-2-7b-chat-hf",
    "meta-llama/Llama-2-13b-chat-hf",
    "mistralai/Mistral-7B-Instruct-v0.1",
    "HuggingFaceH4/zephyr-7b-beta",
    "allenai/tulu-2-7b",
    "allenai/tulu-2-dpo-7b",
}

MODEL_KWARGS = {
    "bigscience/T0pp": {"torch_dtype": torch.bfloat16},
    "bigscience/T0_3B": {"torch_dtype": torch.bfloat16},
    "google/flan-ul2": {"load_in_8bit": True},
    "google/flan-t5-xxl": {"load_in_8bit": True},
    "google/flan-t5-xl": {"load_in_8bit": True},
    "google/flan-t5-large": {"load_in_8bit": True},
    "google/flan-t5-base": {"load_in_8bit": True},
}


TULU_CHAT_TEMPLATE = {
    "chat_template": "{% for message in messages %}\n{% if message['role'] == 'user' %}\n{{ '<|user|>\n' + message['content'] + eos_token }}\n{% elif message['role'] == 'system' %}\n{{ '<|system|>\n' + message['content'] + eos_token }}\n{% elif message['role'] == 'assistant' %}\n{{ '<|assistant|>\n'  + message['content'] + eos_token }}\n{% endif %}\n{% if loop.last and add_generation_prompt %}\n{{ '<|assistant|>' }}\n{% endif %}\n{% endfor %}"
}
TOKENIZER_KWARGS = {
    "allenai/tulu-2-7b": TULU_CHAT_TEMPLATE,
    "allenai/tulu-2-dpo-7b": TULU_CHAT_TEMPLATE,
}


def _is_conversational(model_name_or_path):
    return model_name_or_path in CONVERSATIONAL_MODELS


def _prepare(
    candidates, prompt_file: os.PathLike, model_name_or_path: str, context_file: Optional[os.PathLike] = None
):
    with open(prompt_file) as p:
        prompt_template = "".join(p.readlines()).strip()

    context_passage = None
    if context_file and os.path.exists(context_file):
        with open(context_file, "r") as json_file:
            context_passage = json.load(json_file)

    prompts = []
    for candidate in candidates:
        if isinstance(candidate, Candidate):
            q = candidate.question
            gold_answers = q.answers
            candidate_answer = candidate.answer
            question = q.text
        else:
            gold_answers = candidate["answers"]
            candidate_answer = candidate["candidate_answer"]
            question = candidate["question"]

        gold_answers = ", ".join([f'"{a}"' for a in gold_answers])

        if not question.endswith("?"):
            question += "?"

        if context_passage:
            passage = context_passage.get(question, context_passage.get(question[:-1], None))
            prompt = prompt_template.format(
                q=question, answers=gold_answers, candidate_answer=candidate_answer, passage=passage
            )
        else:
            prompt = prompt_template.format(q=question, answers=gold_answers, candidate_answer=candidate_answer)

        if _is_conversational(model_name_or_path):
            if "Mistral" in model_name_or_path or "tulu-2" in model_name_or_path:
                instruction = None
                content = prompt
            else:
                sections = prompt.split("###")
                instruction = "###".join(sections[:-1]) if len(sections) > 1 else None
                content = sections[-1].strip()

            chat = []
            if instruction:
                chat.append({"role": "system", "content": instruction})

            chat.append({"role": "user", "content": content})
            prompts.append(chat)
        else:
            prompts.append(prompt)

    return prompts


def _prepare_second_pass(chats, model_responses):
    new_chats = []
    for chat, resp in zip(chats, model_responses):
        history = []
        if chat and chat[0]["role"] == "system":
            history.append(chat[1:])
        else:
            history.extend(chat)

        if isinstance(resp, str):
            resp = [resp]

        for r in resp:
            if "###" in r:
                r = r.split("###")[0].strip()

            new_chats.append(
                history
                + [
                    {"role": "assistant", "content": r},
                    {"role": "user", "content": "Please tell me your final judgment in only 'yes' or 'no'"},
                ]
            )

    return new_chats


def _parse_response(response: str, candidate_answer: str, question: str) -> int:
    patterns = [
        r".*['\"]?(yes|no)\.?['\"]?[.!]?$",
        r".*I can answer\s+['\"]?(yes|no)['\"]?[.!]?",
        r".*I would say\s+['\"]?(yes|no)['\"]?[.!]?",
        r".*I must say\s+['\"]?(yes|no)['\"]?[.!]?",
        (r".*my (final )?judgment is\s+['\"]?(yes|no)['\"]?[.!]?", 2),
        r".*I would judge the candidate answer as\s+['\"]?(yes|no)['\"]?[.!]?",
        r".*\s+['\"]?(yes|no)['\"]?,? the candidate( answer)? is",
        r".*[jJ]udgment:\s+['\"]?(yes|no)\.?['\"]?",
    ]
    correct_patterns = [r"candidate( answer)? is correct", r"candidate's correct"]

    if response.lower().startswith("yes"):
        acceptable = "Yes"
    elif response.lower().startswith("no"):
        acceptable = "No"
    else:
        acceptable = ""
        for pattern in patterns:
            match_idx = 1
            if isinstance(pattern, (list, tuple)):
                pattern, match_idx = pattern

            matched = re.match(pattern, response, re.IGNORECASE | re.MULTILINE | re.DOTALL)

            if matched:
                acceptable = matched.group(match_idx).capitalize()
                break
        if not acceptable:
            for pattern in correct_patterns:
                matched = re.search(pattern, response, re.IGNORECASE | re.MULTILINE | re.DOTALL)
                if matched:
                    acceptable = "Yes"
                    break

        if not acceptable:
            logger.warning(f"Invalid response to `{question}` & `{candidate_answer}`: {response}")

    return int(acceptable == "Yes")


def run_inference(
    texts: Union[str, Sequence[str]],
    model,
    tokenizer,
    max_new_tokens: int = 256,
    do_sample: bool = True,
    top_p: float = 1.0,
    num_beams: int = 1,
    batch_size: int = 1,
    num_workers: int = 16,
    num_return_sequences: int = 1,
):
    if isinstance(texts, str):
        texts = [texts]
    device = "cuda" if torch.cuda.is_available() else ("mps" if torch.backends.mps.is_available() else "cpu")

    model.eval()

    dataset = datasets.Dataset.from_list(
        [
            {
                "text": (
                    tokenizer.apply_chat_template(t, tokenize=False, add_generation_prompt=True)
                    if _is_conversational(model.config.name_or_path)
                    else t
                )
            }
            for t in texts
        ]
    )

    dataset = dataset.map(
        lambda sample: tokenizer(sample["text"]),
        batched=True,
        remove_columns=list(dataset.features),
    )

    test_dataloader = DataLoader(
        dataset,
        batch_size=batch_size,
        num_workers=num_workers,
        pin_memory=True,
        collate_fn=DataCollatorForSeq2Seq(tokenizer, padding="longest")
        if model.config.is_encoder_decoder
        else DataCollatorWithPadding(tokenizer, padding="longest"),
    )

    outputs = []
    for batch in tqdm(test_dataloader):
        for key in batch.keys():
            batch[key] = batch[key].to(device)

        batch_size, seq_length = batch["input_ids"].shape

        with torch.no_grad():
            output = model.generate(
                **batch,
                do_sample=do_sample,
                max_new_tokens=max_new_tokens,
                top_p=top_p,
                num_beams=num_beams,
                num_return_sequences=num_return_sequences,
            )

        output = output.reshape(batch_size, num_return_sequences, -1)

        for b in range(batch_size):
            output_ids = output[b]

            _outs = []
            for s in range(num_return_sequences):
                _ids = output_ids[s]
                if not model.config.is_encoder_decoder:
                    _ids = _ids[seq_length:]

                _outs.append(ftfy.fix_encoding(tokenizer.decode(_ids, skip_special_tokens=True).strip()))

            outputs.append(_outs[0] if len(_outs) == 1 else _outs)

    return outputs


def llm_eval(model_name_or_path: str, candidates, **kwargs):
    prompt_file = kwargs.pop("prompt_file", None)
    context_file = kwargs.pop("context_file", None)
    num_return_sequences = kwargs.pop("num_return_sequences", 1)

    assert prompt_file and os.path.exists(prompt_file), "prompt_file is required in llm_eval"

    examples = _prepare(candidates, prompt_file, model_name_or_path, context_file)
    
    # Ollama API로 프롬프트 전송 및 응답 수집
    responses = []
    for prompt in tqdm(examples):
        # chat 형식이면 마지막 user 메시지만 추출
        if isinstance(prompt, list):
            if prompt[-1]["role"] == "user":
                prompt_text = prompt[-1]["content"]
            else:
                prompt_text = prompt[-1]
        else:
            prompt_text = prompt
        try:
            resp = requests.post(
                "http://localhost:11434/api/generate",
                json={"model": model_name_or_path, "prompt": prompt_text, "stream": False}
            )
            result = resp.json()
            responses.append(result["response"])
        except Exception as e:
            logger.warning(f"Ollama API 호출 실패: {e}")
            responses.append("")
    original_responses = responses

    # 기존 LLM 평가 방식과 동일하게 파싱
    outputs = []
    for index in range(len(candidates)):
        num_judgments = 1
        judgments = []
        resp = responses[index]
        judgments.append(_parse_response(resp, candidates[index].answer, candidates[index].question.text))
        acceptable_count = sum(judgments)
        outputs.append((round(acceptable_count / num_judgments), original_responses[index], judgments))

    return outputs
